torch/csrc/jit/codegen/cuda/root_domain_map.cpp

```diff
@@ -0,0 +1,379 @@
+  for (const auto& chain: use_chains) {
+    for (const auto& tv: ir_utils::filterByType<TensorView>(chain)) {
+      const auto& root_domain = tv->getRootDomain();
+      for (const auto& id: root_domain) {
+        DomainKey consumer_key(tv->domain(), id);
+        for (const auto& reduction_key: reduction_keys) {
+          inconsistent_domains_.at(reduction_key).insert(consumer_key);
+        }
+      }
+    }
+  }
+}
+
+bool FindIncompatibleDomains::isReductionOutputMerged(
+    const std::vector<DomainKey>& consumer_domains,
+    const DisjointSet<DomainKey, DomainKeyHash>& eq_set) const {
+  for (const auto& kv: inconsistent_domains_) {
+    const auto& reducion_domain = kv.first;
+    const auto& incompatible_domains = kv.second;
+    //std::cerr << "Inconsistent reduction: " << reducion_domain << std::endl;
+    DomainKey consumer_domain_with_reduction;
+    bool reduction_found = false;
+    for (const auto& consumer_domain: consumer_domains) {
+      if (eq_set.areEquivalent(reducion_domain, consumer_domain)) {
+        consumer_domain_with_reduction = consumer_domain;
+        //std::cerr << "reduction found in: " << consumer_domain << std::endl;
+        reduction_found = true;
+        break;
+      }
+    }
+    if (!reduction_found) {
+      //std::cerr << "reduction not found in consumers\n";
+      continue;
+    }
+    // Make sure no incompatible domains will be merged with the reduction domain.
+    for (const auto& consumer_domain: consumer_domains) {
+      if (consumer_domain == consumer_domain_with_reduction) {
+        continue;
+      }
+      if (std::any_of(incompatible_domains.begin(), incompatible_domains.end(),
+                      [&](const DomainKey& incompatible_domain) {
+                        return eq_set.areEquivalent(consumer_domain, incompatible_domain);
+                      })) {
+        // Merging them will result in inconsistency
+        //std::cerr << "incompatible consumer found: " << consumer_domain << std::endl;
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+RootDomainMap::RootDomainMap() {
+  Fusion* fusion = FusionGuard::getCurFusion();
+  TORCH_INTERNAL_ASSERT(fusion != nullptr);
+  // std::cerr << "Traversing for root domain map\n";
+  std::vector<Val*> leaves;
+  for (Val* val : fusion->deterministic_vals()) {
+    if (!fusion->used(val)) {
+      leaves.push_back(val);
+    }
+    // Register all domain keys. This simplifies the matching analysis.
+    if (val->isA<TensorView>()) {
+      auto td = val->as<TensorView>()->domain();
+      for (const auto& id: td->getRootDomain()) {
+        eq_set_.insert(DomainKey(td, id));
+      }
+      if (td->hasRFactor()) {
+        for (const auto& id: td->getRFactorDomain()) {
+          eq_set_.insert(DomainKey(td, id));
+        }
+      }
+    }
+  }
+  traverseFrom(fusion, leaves, false);
+  //std::cerr << "Traversal done\n";
+  //print(std::cerr);
+  if (!pending_map_.empty()) {
```

Comment:
Traversal of `Expr`s adds entries into `pending_map_`, visiting `TensorView`s will process those pending entries.